### PR TITLE
LaTeX: add latex_docclassprefix configuration setting

### LINKED ISSUE #6672: Remove chapters for custom howto LaTeX classes
**Is your feature request related to a problem? Please describe.**
Since the `article` LaTeX class, which the Sphinx `howto` class is based upon, does not support chapters they are removed according to the following control structure in `sphinx/writers/latex.py`:

        # we assume LaTeX class provides \chapter command except in case
        # of non-Japanese 'howto' case
        self.sectionnames = LATEXSECTIONNAMES[:]
        if self.settings.docclass == 'howto':
            docclass = self.config.latex_docclass.get('howto', 'article')
            if docclass[0] == 'j':  # Japanese class...
                pass
            else:
                self.sectionnames.remove('chapter')
        else:
            docclass = self.config.latex_docclass.get('manual', 'report')
        self.elements['docclass'] = docclass

We use custom LaTeX classes that are based on (i.e. `\LoadClass` them) the default Sphinx `howto` and `manual` classes. However, it is not possible to trigger the removal of chapters with a custom class name. The option `latex_toplevel_sectioning` only works on a global level and does not allow us to mix `howto`/`manual`-style classes in the `latex_documents` list anymore.

**Describe the solution you'd like**
An easy solution would be to replace the line

        if self.settings.docclass == 'howto':

by

        if 'howto' in self.settings.docclass:

enabling the removal of chapters for any class containing the keyword `howto`.

**Describe alternatives you've considered**
Have not come up with any other solution so far.

### PR DESCRIPTION
**Pull request by jfbu**
Closes: #6672

### Feature or Bugfix
- Feature

### CODE REVIEW

File: doc/latex.rst [hunk 1]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+The LaTeX document class
+------------------------
+
+LaTeX document classes are files (with extension ``.cls``) containing LaTeX
+macros, often devoted to custom layout definitions.  They are loaded via
+the ``\documentclass{<classname>}`` line in a LaTeX source file.
+
+Sphinx provides two LaTeX classes ``sphinxhowto`` and ``sphinxmanual`` of its
+own (the two ``.cls`` files are automatically copied over to the LaTeX build
+repertory).  Which one is used depends on whether ``'howto'`` or ``'manual'``
```
Comment (sephalon) [inline doc/latex.rst]: "repertory" should probably be "directory"

File: doc/latex.rst [hunk 2]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+is encountered in the :confval:`latex_documents` variable (the default uses
+``'manual'``).  Any string distinct from these two is interpreted as the name
+of a LaTeX class which will be loaded in place of the Sphinx custom ones.  In
+such case the ``\sphinxmaketitle``, ``\sphinxtableofcontents``,
```
Comment (sephalon) [inline doc/latex.rst]: In this case, ``sphinx.sty`` provides fallbacks for ``\sphinxmaketitle``, ``\sphinxtableofcontents``, ``sphinxthebibliography`` and ``sphinxtheindex`` to the default LaTeX commands and environments without ``sphinx`` prefix.

File: doc/latex.rst [hunk 3]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+``sphinxthebibliography``, and ``sphinxtheindex`` fall back to the meaning of
+the LaTeX macros and environments without ``sphinx`` prefix.
+
+The Sphinx LaTeX classes are slim wrappers of standard classes: respectively
```
Comment (sephalon) [inline doc/latex.rst]: The Sphinx classes ``sphinxhowto`` and ``sphinxmanual`` are slim wrappers around the standard LaTeX classes ``article`` and ``report`` respectively (Japanese documents use ``jreport`` and ``jsbook`` instead).

File: doc/latex.rst [hunk 4]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+the ``article`` and ``report`` ones (Japanese documents use ``jreport`` and
+``jsbook``).  It is possible to modify these base classes via the
```
Comment (sephalon) [inline doc/latex.rst]: It is possible to alter the mapping between Sphinx and base classes via the :confval:`latex_docclass` dictionary.

File: doc/latex.rst [hunk 5]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+With :confval:`latex_docclassprefix` it is even possible to replace usage of
+``sphinxhowto.cls`` or ``sphinxmanual.cls`` by classes of one's own, while
+keeping the choice of either ``'howto'`` or ``'manual'`` (which, apart for
+Japanese documents, means whether or not the ``\chapter`` macro is used in the
+LaTeX mark-up). For example they will be ``myhowto.cls`` or ``mymanual.cls``
+if the variable value is ``'my'``.  This is useful if one wants to use one of
+the two Sphinx classes but override some of its settings: one only needs then
+to define a custom class loading the Sphinx one via ``\LoadClass``.  The
+:confval:`latex_additional_files` will be useful to get the custom class files
+be copied over to the LaTeX build repertory (this is for portability of a
```
Comment (sephalon) [inline doc/latex.rst]: "repertory" should probably be "dictionary", see also line 79

File: doc/latex.rst [hunk 6]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+such case the ``\sphinxmaketitle``, ``\sphinxtableofcontents``,
+``sphinxthebibliography``, and ``sphinxtheindex`` fall back to the meaning of
+the LaTeX macros and environments without ``sphinx`` prefix.
+
+The Sphinx LaTeX classes are slim wrappers of standard classes: respectively
+the ``article`` and ``report`` ones (Japanese documents use ``jreport`` and
+``jsbook``).  It is possible to modify these base classes via the
+:confval:`latex_docclass` dictionary.
+
+With :confval:`latex_docclassprefix` it is even possible to replace usage of
```
Comment (sephalon) [inline doc/latex.rst]: Remove "usage of"

File: doc/latex.rst [hunk 7]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+``sphinxhowto.cls`` or ``sphinxmanual.cls`` by classes of one's own, while
+keeping the choice of either ``'howto'`` or ``'manual'`` (which, apart for
+Japanese documents, means whether or not the ``\chapter`` macro is used in the
```
Comment (sephalon) [inline doc/latex.rst]: Replace "means" by "decides"

File: doc/latex.rst [hunk 8]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+LaTeX mark-up). For example they will be ``myhowto.cls`` or ``mymanual.cls``
```
Comment (sephalon) [inline doc/latex.rst]: For example, the class ``myhowto.cls`` or ``mymanual.cls`` respectively will be loaded if :confval:`latex_docclassprefix` is set to ``my``.

File: doc/usage/configuration.rst [hunk 9]
```diff
@@ -2023,6 +2029,14 @@ These options influence LaTeX output.
       In Japanese docs (:confval:`language` is ``'ja'``), by default
       ``'jreport'`` is used for ``'howto'`` and ``'jsbook'`` for ``'manual'``.
 
+.. confval:: latex_docclassprefix
+
+   A string (defaults to ``'sphinx'``) which serves as prefix to map
```
Comment (sephalon) [inline doc/usage/configuration.rst]: as a prefix

File: doc/latex.rst [hunk 10]
```diff
@@ -65,6 +65,43 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+the LaTeX macros and environments without ``sphinx`` prefix.
+
+The Sphinx LaTeX classes are slim wrappers of standard classes: respectively
+the ``article`` and ``report`` ones (Japanese documents use ``jreport`` and
+``jsbook``).  It is possible to modify these base classes via the
+:confval:`latex_docclass` dictionary.
+
+With :confval:`latex_docclassprefix` it is even possible to replace usage of
+``sphinxhowto.cls`` or ``sphinxmanual.cls`` by classes of one's own, while
+keeping the choice of either ``'howto'`` or ``'manual'`` (which, apart for
```
Comment (sephalon) [inline doc/latex.rst]: We should maybe emphasize here that ``howto`` and ``manual`` are special keywords (chapter pruning, Japanese language), while any other class name can be used as well.

**Review by sephalon — COMMENTED** [review]
Great, that should clarify the situation. I added some comments on the documentation.

File: doc/latex.rst (hunk 6 above)
Comment (jfbu) [reply doc/latex.rst]: This become ambiguous. The user-prefixed ``howto`` and ``manual`` classes will not replaced the Sphinx ones. The Sphinx ones are still copied over to the LaTeX build directory. The user-prefixed ones replace them in terms of their usage, not in terms of their presence in the build directory.

**Comment by jfbu** [comment]
@sephalon thanks a lot for reviewing. Please tell me what you think of current version.

File: doc/latex.rst [hunk 11]
```diff
@@ -65,6 +65,47 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+
+Sphinx provides two LaTeX classes ``sphinxhowto`` and ``sphinxmanual`` of its
+own (the two ``.cls`` files are automatically copied over to the LaTeX build
+directory).  Which one is used depends on whether ``'howto'`` or ``'manual'``
+is encountered in the :confval:`latex_documents` variable (the default uses
+``'manual'``).  Any string distinct from these two is interpreted as the name
+of a LaTeX class which will be loaded in place of the Sphinx custom ones.  In
+this case, ``sphinx.sty`` will let ``\sphinxmaketitle``,
+``\sphinxtableofcontents``, ``sphinxthebibliography`` and ``sphinxtheindex``
+default to the LaTeX commands and environments without sphinx prefix.
```
Comment (sephalon) [inline doc/latex.rst]: Either capitalize "sphinx" or put in backticks

File: doc/latex.rst [hunk 12]
```diff
@@ -65,6 +65,47 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+particular with ``'howto'`` the Sphinx LaTeX writer is told not to use the
+``\chapter`` mark-up (except for Japanese).  It is possible to combine their
+usage with custom classes.  This is done with the help of
+:confval:`latex_docclassprefix`.  If for example it is set to ``'my'`` then
+``myhowto.cls`` or ``mymanual.cls`` respectively will be loaded.  The user is
+free inside such a custom class to ``\LoadClass`` either ``sphinxhowto`` or
+``sphinxmanual``.  This allows to keep some of their custom settings and also
+override others.  This way, the base classes as specified via
+:confval:`latex_docclass` will get loaded as usual; else it is the
+responsability of the custom class to load if needed such a base class.
```
Comment (sephalon) [inline doc/latex.rst]: responsibility

File: doc/latex.rst [hunk 13]
```diff
@@ -65,6 +65,47 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+default to the LaTeX commands and environments without sphinx prefix.
+
+The Sphinx classes are slim wrappers around the standard LaTeX classes
+``article`` and ``report`` respectively (Japanese documents use ``jreport``
+and ``jsbook`` instead).  It is possible to alter the mapping between Sphinx
+and base classes via the :confval:`latex_docclass` dictionary.
+
+As explained above, ``'howto'`` and ``'manual'`` are special keywords, and in
+particular with ``'howto'`` the Sphinx LaTeX writer is told not to use the
+``\chapter`` mark-up (except for Japanese).  It is possible to combine their
```
Comment (sephalon) [inline doc/latex.rst]: With :confval:`latex_docclassprefix`, one can leverage this special treatment of ``howto`` and ``manual`` in custom classes.

File: doc/latex.rst (hunk 12 above)
Comment (sephalon) [inline doc/latex.rst]: …to load such a base class if necessary.

File: doc/latex.rst [hunk 14]
```diff
@@ -65,6 +65,47 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+usage with custom classes.  This is done with the help of
+:confval:`latex_docclassprefix`.  If for example it is set to ``'my'`` then
+``myhowto.cls`` or ``mymanual.cls`` respectively will be loaded.  The user is
+free inside such a custom class to ``\LoadClass`` either ``sphinxhowto`` or
+``sphinxmanual``.  This allows to keep some of their custom settings and also
```
Comment (sephalon) [inline doc/latex.rst]: …to ``\LoadClass`` either ``sphinxhowto`` or ``sphinxmanual`` in order to keep some of their custom settings and override others.

File: doc/latex.rst [hunk 15]
```diff
@@ -65,6 +65,47 @@ avoid interpretation as escape sequences, or use raw strings (as is done here).
...
+The Sphinx classes are slim wrappers around the standard LaTeX classes
+``article`` and ``report`` respectively (Japanese documents use ``jreport``
+and ``jsbook`` instead).  It is possible to alter the mapping between Sphinx
+and base classes via the :confval:`latex_docclass` dictionary.
+
+As explained above, ``'howto'`` and ``'manual'`` are special keywords, and in
+particular with ``'howto'`` the Sphinx LaTeX writer is told not to use the
+``\chapter`` mark-up (except for Japanese).  It is possible to combine their
+usage with custom classes.  This is done with the help of
+:confval:`latex_docclassprefix`.  If for example it is set to ``'my'`` then
```
Comment (sephalon) [inline doc/latex.rst]: If, for example, :confval:`latex_docclassprefix` is set to… (avoid ambiguity of "it" in technical documentation)

**Review by sephalon — COMMENTED** [review]
Few more comments. Otherwise LGTM!

**Comment by jfbu** [comment]
Renewed thanks @sephalon for your further helpful improvements. I have merged them with some slight reformulation at places. Tell me what you think when you get the time.

**Comment by sephalon** [comment]
Very clear now. Thanks for implementing this feature so promptly!

**Comment by jfbu** [comment]
Thanks a lot @sephalon.

Are you ok with this added configuration @tk0miya?

**Comment by tk0miya** [comment]
Sorry for late response.

IMO, it is better to describe the behavior of custom docclass like `theme.conf` for HTML themes. In #6672's case, it is needed to choose toplevel sectioning of the docclass:
```
# theme.conf for LaTeX docclass
[theme]
toplevel_sectioning = chapter
```

**Comment by jfbu** [comment]
@tk0miya thanks for comment but I don't understand it, we don't have themes for LaTeX... and toplevel in LaTeX can still be `\part` even if `\chapter` is not used.

**Comment by tk0miya** [comment]
I made #6969 an alternative of this. So closing this now.

**Pull request closed without merging**